File: internal/compute/pipeline.go
package compute

import (
	"pii/internal/compute/detect"
	"pii/internal/compute/mask"
)

// Pipeline orchestrates detect → context → merge → mask.
type Pipeline struct {
	structural *detect.Registry
	ner        *detect.NERDetector
	context    *detect.ContextRule
	masker     *mask.Masker
}

// NewPipeline builds a detection pipeline.
func NewPipeline(structural *detect.Registry, ner *detect.NERDetector, context *detect.ContextRule, masker *mask.Masker) *Pipeline {
	return &Pipeline{
		structural: structural,
		ner:        ner,
		context:    context,
		masker:     masker,
	}
}

// Process masks text and returns the masked text and detected types.
func (p *Pipeline) Process(text string) (string, []string) {
	return p.ProcessFiltered(text, nil)
}

// ProcessFiltered masks text, keeping only spans whose type is allowed.
// A nil or empty allowed set keeps every detected type.
func (p *Pipeline) ProcessFiltered(text string, allowed map[string]bool) (string, []string) {
	spans := p.structural.DetectAll(text)
	nerSpans := p.ner.Detect(text)
	spans = append(spans, p.context.Filter(nerSpans, text)...)
	merged := detect.MergeSpans(spans)

	if len(allowed) > 0 {
		filtered := merged[:0]
		for _, s := range merged {
			if allowed[s.Type] {
				filtered = append(filtered, s)
			}
		}
		merged = filtered
	}

	masked := p.masker.Apply(text, merged)

	seen := map[string]struct{}{}
	var types []string
	for _, s := range merged {
		if _, ok := seen[s.Type]; ok {
			continue
		}
		seen[s.Type] = struct{}{}
		types = append(types, s.Type)
	}
	return masked, types
}

File: compute/src/pipeline.rs
//! Конвейер обработки: детекция → контекст → слияние → маскирование.

use crate::detect::base::DetectorRegistry;
use crate::detect::context::ContextRule;
use crate::detect::ner::NERDetector;
use crate::mask::masker::Masker;

/// Конвейер обработки текста.
pub struct DetectionPipeline {
    /// Реестр структурных детекторов.
    structural: DetectorRegistry,
    /// NER-детектор.
    ner: NERDetector,
    /// Контекстное правило.
    context: ContextRule,
    /// Маскер.
    masker: Masker,
}

impl DetectionPipeline {
    /// Обрабатывает текст: возвращает маскированный текст и типы ПД.
    pub async fn process(&self, text: &str) -> (String, Vec<String>) {
        todo!("реализация конвейера обработки")
    }
}